agent/consul/leader_connect_ca.go

```diff
@@ -0,0 +1,1082 @@
+		intermediateCert.NotAfter) {
+		return nil
+	}
+
+	// Enough time has passed, go ahead with getting a new intermediate.
+	renewalFunc := c.getIntermediateCAPrimary
+	if !isPrimary {
+		renewalFunc = c.getIntermediateCASigned
+	}
+	if err := renewalFunc(provider, activeRoot); err != nil {
+		return err
+	}
+
+	if err := c.persistNewRootAndConfig(provider, activeRoot, nil); err != nil {
+		return err
+	}
+
+	c.setCAProvider(provider, activeRoot)
+	return nil
+}
+
+// secondaryCARootWatch maintains a blocking query to the primary datacenter's
+// ConnectCA.Roots endpoint to monitor when it needs to request a new signed
+// intermediate certificate.
+func (c *CAManager) secondaryCARootWatch(ctx context.Context) error {
+	args := structs.DCSpecificRequest{
+		Datacenter: c.srv.config.PrimaryDatacenter,
+		QueryOptions: structs.QueryOptions{
+			// the maximum time the primary roots watch query can block before returning
+			MaxQueryTime: c.srv.config.MaxQueryTime,
+		},
+	}
+
+	c.logger.Debug("starting Connect CA root replication from primary datacenter", "primary", c.srv.config.PrimaryDatacenter)
+
+	retryLoopBackoff(ctx, func() error {
+		var roots structs.IndexedCARoots
+		if err := c.srv.forwardDC("ConnectCA.Roots", c.srv.config.PrimaryDatacenter, &args, &roots); err != nil {
+			return fmt.Errorf("Error retrieving the primary datacenter's roots: %v", err)
+		}
+
+		// Attempt to update the roots using the returned data.
+		if err := c.UpdateRoots(roots); err != nil {
+			return err
+		}
+		args.QueryOptions.MinQueryIndex = nextIndexVal(args.QueryOptions.MinQueryIndex, roots.QueryMeta.Index)
+		return nil
+	}, func(err error) {
+		c.logger.Error("CA root replication failed, will retry",
+			"routine", secondaryCARootWatchRoutineName,
+			"error", err,
+		)
+	})
+
+	return nil
+}
+
+// UpdateRoots updates the cached roots from the primary and regenerates the intermediate
+// certificate if necessary.
+func (c *CAManager) UpdateRoots(roots structs.IndexedCARoots) error {
+	// Update the state first to claim the 'lock'.
+	if err := c.setState(CAStateReconfig, true); err != nil {
+		return err
+	}
+	defer c.setState(CAStateReady, false)
+
+	// Update the cached primary roots now that the lock is held.
+	if err := c.setPrimaryRoots(roots); err != nil {
+		return err
+	}
+
+	// Check to see if the primary has been upgraded in case we're waiting to switch to
+	// secondary mode.
+	provider, _ := c.getCAProvider()
+	if provider == nil {
+		// this happens when leadership is being revoked and this go routine will be stopped
+		return nil
+	}
+	if !c.configuredSecondaryCA() {
+		versionOk, primaryFound := ServersInDCMeetMinimumVersion(c.srv, c.srv.config.PrimaryDatacenter, minMultiDCConnectVersion)
+		if !primaryFound {
+			return fmt.Errorf("Primary datacenter is unreachable - deferring secondary CA initialization")
+		}
+
+		if versionOk {
+			if err := c.initializeSecondaryProvider(provider, roots); err != nil {
```

Comment:
if this errors, what mechanism attempts to re-execute it? We already updated the stored roots above

Reply:
It won't be reattempted currently until the next run of the root watch routine returns, Do you think we need a separate retry backoff loop here even when this function (UpdateRoots) is being called periodically?

Reply:
I think as long as the system does something reasonable to recover from a temporary failure here that would be sufficient.